controllers/common.go

```diff
@@ -119,18 +119,32 @@ func BuildTargetAllocator(params targetallocator.Params) ([]client.Object, error
 // getList queries the Kubernetes API to list the requested resource, setting the list l of type T.
 func getList[T client.Object](ctx context.Context, cl client.Client, l T, options ...client.ListOption) (map[types.UID]client.Object, error) {
 	ownedObjects := map[types.UID]client.Object{}
-	list := &unstructured.UnstructuredList{}
 	gvk, err := apiutil.GVKForObject(l, cl.Scheme())
 	if err != nil {
 		return nil, err
 	}
-	list.SetGroupVersionKind(gvk)
-	err = cl.List(ctx, list, options...)
+	gvk.Kind = fmt.Sprintf("%sList", gvk.Kind)
```

Comment:
This feels really stupid, but I couldn't find an idiomatic way to get a `xxxList` instance given an `xxx` instance.

Reply:
FYI, if I go back to the previous version using `Unstructured`, then the cache doesn't work even if I explicitly enable unstructured caching in the client. I think I'd need to change the indexer as well.

Reply:
For what's it's worth, controller-runtime itself does these kinds of things as well: https://github.com/kubernetes-sigs/controller-runtime/blob/f529320b4c0fa5205067dec4eaeabc3c6cf62d25/pkg/client/client.go#L230.

And if the object doesn't support `List`, well, the test won't pass. We could even add a test with a fake client to verify.

I would like a solution without these kinds of hacks, but I'm losing hope that it's actually possible.

Reply:
yeah it's a huge bummer.